### PR TITLE
feat(payment): PAYPAL-1630 BraintreePayPal on PDP

### PR DESCRIPTION
**Pull request by bc-nick**
## What?
Quick payment button for BraintreePaypal provider

## Why?
According to task [PAYPAL-1630](https://bigcommercecloud.atlassian.net/browse/PAYPAL-1630)

## Testing / Proof

https://user-images.githubusercontent.com/99336044/194883965-de1cca2d-9d31-4515-8445-cfabaf18f82d.mov

@bigcommerce/checkout @bigcommerce/payments

### CODE REVIEW

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 1]
```diff
@@ -188,6 +250,7 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
                 device_data: deviceData,
                 billing_address: JSON.stringify(mapToLegacyBillingAddress(details)),
                 shipping_address: JSON.stringify(mapToLegacyShippingAddress(details)),
+                ...this._buyNowCartData?.id && { cart_id: this._buyNowCartData?.id },
```
Comment (serhii-tkachenko) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:253]: It might be better to create a variable for `this._buyNowCartData?.id`

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 2]
```diff
@@ -1,22 +1,29 @@
...
-import { CheckoutActionCreator, CheckoutStore } from '../../../checkout';
+import { CartRequestSender, Cart } from '../../../cart';
+import { BuyNowCartCreationError } from '../../../cart/errors';
+import { CheckoutActionCreator, CheckoutStore, InternalCheckoutSelectors } from '../../../checkout';
 import { InvalidArgumentError, MissingDataError, MissingDataErrorType, StandardError } from '../../../common/error/errors';
 import { mapToBraintreeShippingAddressOverride, BraintreeError, BraintreePaypalCheckout, BraintreeSDKCreator, BraintreeTokenizePayload } from '../../../payment/strategies/braintree';
 import { PaypalAuthorizeData, PaypalHostWindow } from '../../../payment/strategies/paypal';
 import { CheckoutButtonInitializeOptions } from '../../checkout-button-options';
 import CheckoutButtonStrategy from '../checkout-button-strategy';
+import PaymentMethod from '../../../payment/payment-method';
```
Comment (serhii-tkachenko) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:11]: It might be better to place this line on line #7

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 3]
```diff
@@ -38,9 +45,24 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
...
+
+        if (braintreepaypal?.buyNowInitializeOptions) {
+            state = this._store.getState();
+            paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
+
+            if (!braintreepaypal.currencyCode) {
+                throw new InvalidArgumentError(`Unable to initialize payment because "options.braintreepaypal.currency" argument is not provided.`);
+            }
+
+            currency = braintreepaypal.currencyCode;
```
Comment (serhii-tkachenko) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:60]:
```suggestion
            currencyCode = braintreepaypal.currencyCode;
```

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 4]
```diff
@@ -38,9 +45,24 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
...
+        } else {
+            state = await this._store.dispatch(this._checkoutActionCreator.loadDefaultCheckout());
+            paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
+            currency = state.cart.getCartOrThrow()?.currency.code;
```
Comment (serhii-tkachenko) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:64]:
```suggestion
            currencyCode = state.cart.getCartOrThrow()?.currency.code;
```

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 5]
```diff
@@ -38,9 +45,24 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
             throw new InvalidArgumentError(`Unable to initialize payment because "options.braintreepaypal" argument is not provided.`);
         }
 
-        const state = await this._store.dispatch(this._checkoutActionCreator.loadDefaultCheckout());
-        const paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
-        const currency = state.cart.getCartOrThrow()?.currency.code;
+        let state: InternalCheckoutSelectors;
+        let paymentMethod: PaymentMethod<any>;
+        let currency: string;
```
Comment (serhii-tkachenko) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:50]:
```suggestion
        let currencyCode: string;
```

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 6]
```diff
@@ -120,6 +142,8 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
     }
 
     private _renderPayPalMessages(messagingContainerId?: string): void {
+        if (!messagingContainerId) return;
```
Comment (serhii-tkachenko) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:145]: We already have this check in this method. It might be better to place state and cart variables inside IF block.

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 7]
```diff
@@ -137,16 +161,25 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
...
 
         try {
-            const checkout = state.checkout.getCheckoutOrThrow();
-            const config = state.config.getStoreConfigOrThrow();
+            this._buyNowCart = await this._createBuyNowCart({ buyNowInitializeOptions });
+
+            if (this._buyNowCart) {
+                state = this._store.getState();
+            } else {
+                state = await this._store.dispatch(this._checkoutActionCreator.loadDefaultCheckout());
```
Comment (serhii-tkachenko) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:175]: Do we still need to load default checkout here?

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 8]
```diff
@@ -155,19 +188,37 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
...
-                currency: config.currency.code,
+                amount,
+                currency: currencyCode,
                 offerCredit: false,
             });
         } catch (error) {
-            if (onError) {
-                onError(error);
+            if (onPaymentError) {
+                onPaymentError(error);
```
Comment (serhii-tkachenko) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:197]: Why did you change this one? As far as I see we are creating payment but not submitting it.
Please double check the description for the callback that you use:
https://github.com/bigcommerce/checkout-sdk-js/pull/1629/files#diff-718de5b34e6160523a486f1c1c5702d7449591d7ad1664d4e3ce1d7a79072289R38

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 9]
```diff
@@ -179,6 +230,7 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
             const { deviceData } = await this._braintreeSDKCreator.getDataCollector({ paypal: true });
             const tokenizePayload = await braintreePaypalCheckout.tokenizePayment(authorizeData);
             const { details, nonce } = tokenizePayload;
+            const cart_id = this._buyNowCart?.id;
```
Comment (serhii-tkachenko) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:233]:
```suggestion
            const buyNowCartId = this._buyNowCart?.id;
```

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 10]
```diff
@@ -188,6 +240,7 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
                 device_data: deviceData,
                 billing_address: JSON.stringify(mapToLegacyBillingAddress(details)),
                 shipping_address: JSON.stringify(mapToLegacyShippingAddress(details)),
+                ...cart_id && { cart_id },
```
Comment (serhii-tkachenko) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:243]:
```suggestion
                ... buyNowCartId && { cart_id: buyNowCartId },
```

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts (hunk 9 above)
Comment (serhii-tkachenko) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:233]: We are not using snake case for variables name

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts (hunk 8 above)
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:197]: `onPaymentError` it is `onError` from `braintreepaypal`
I removed `onError` that we provide to `_setupPayment` and do destructurization there

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts (hunk 7 above)
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:175]: Yes
This is required because we need get checkout for the non 'Buy Now' logic

Comment (serhii-tkachenko) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:175]: But we are getting it in the initialization method. Why should we repeat the loading here?

Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:175]: @serhii-tkachenko We can create a new private variable for **state** like `private _state: InternalCheckoutSelectors` and get it from there. How do you think?

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 11]
```diff
@@ -120,11 +142,13 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
     }
 
     private _renderPayPalMessages(messagingContainerId?: string): void {
+        const isMessageContainerAvailable = messagingContainerId && Boolean(document.getElementById(messagingContainerId));
+
+        if (!isMessageContainerAvailable) return;
+
```
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts]: Good idea. Thanks

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts (hunk 7 above)
Comment (serhii-tkachenko) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:172]: It is not necessary due to `this._store.getState()` method, because the state is already has all data that we need. I think that it is a question of another PR, lets skip it for now

**Review by serhii-tkachenko — APPROVED** [review]
LGTM now 👍

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 12]
```diff
@@ -38,15 +45,30 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
             throw new InvalidArgumentError(`Unable to initialize payment because "options.braintreepaypal" argument is not provided.`);
         }
 
-        const state = await this._store.dispatch(this._checkoutActionCreator.loadDefaultCheckout());
-        const paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
-        const currency = state.cart.getCartOrThrow()?.currency.code;
+        let state: InternalCheckoutSelectors;
+        let paymentMethod: PaymentMethod<any>;
```
Comment (animesh1987) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:49]: Can we avoid the use of `any`?

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 13]
```diff
@@ -38,15 +45,30 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
...
+            paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
+            currencyCode = state.cart.getCartOrThrow()?.currency.code;
+        }
 
         if (!paymentMethod.clientToken) {
             throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
         }
 
-        const paypalCheckoutOptions = { currency };
+        const paypalCheckoutOptions = { currency: currencyCode };
```
Comment (animesh1987) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:71]: Is this variable used anywhere in this method?

Comment (animesh1987) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:71]: Found the use, ignore it.

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts (hunk 12 above)
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts]: Yes
It is default generic type for PaymentMethod
I removed it, thanks

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 14]
```diff
@@ -137,16 +166,27 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
...
     private async _setupPayment(
         braintreePaypalCheckout: BraintreePaypalCheckout,
+        braintreepaypal: BraintreePaypalButtonInitializeOptions,
         shippingAddress?: Address | null,
-        onError?: (error: BraintreeError | StandardError) => void
+        onError?: (error: BraintreeError | StandardError) => void,
     ): Promise<string> {
-        const state = await this._store.dispatch(this._checkoutActionCreator.loadDefaultCheckout());
+        const { buyNowInitializeOptions, currencyCode } = braintreepaypal;
+        let state: InternalCheckoutSelectors;
```
[the remark this answers is not in the document]
Comment (animesh1987) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:164]: Can you address this please

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 15]
```diff
@@ -137,16 +158,25 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
...
+            if (this._buyNowCart) {
+                state = this._store.getState();
+            } else {
+                state = await this._store.dispatch(this._checkoutActionCreator.loadDefaultCheckout());
+            }
+
             const customer = state.customer.getCustomer();
 
+            const amount = this._buyNowCart ? this._buyNowCart?.cartAmount : state.checkout.getCheckoutOrThrow().outstandingBalance;
+            const currencyCode = this._buyNowCart ? braintreepaypal.currencyCode : state.config.getStoreConfigOrThrow().currency.code;
```
Comment (animesh1987) [inline packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:178]: Don't we validate above for `currencyCode` to be in `braintreepaypal` object?

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts [hunk 16]
```diff
@@ -137,16 +166,27 @@ export default class BraintreePaypalButtonStrategy implements CheckoutButtonStra
...
+                state = this._store.getState();
+            } else {
+                state = await this._loadDefaultCheckout();
+            }
+
             const customer = state.customer.getCustomer();
 
+            const amount = this._buyNowCartData ? this._buyNowCartData?.baseAmount : state.checkout.getCheckoutOrThrow().outstandingBalance;
+            const currency = this._buyNowCartData ? currencyCode : state.config.getStoreConfigOrThrow().currency.code;
+
```
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts]: For both methods this is not possible because we don't have `checkout` on process `BuyNow` flow

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:164]: @serhii-tkachenko It does not fit this case because state variable rewrite below

File: packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts (hunk 15 above)
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:178]: @animesh1987 Yeap
We validate it in `initialize` method for if this is BuyNow flow
But here we need to understand where we must get this value

Comment (animesh1987) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts:178]: if the value is validated above, that should mean that we use the value in `braintreepaypal ` and don't need to have a fallback on display currency.

Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/braintree/braintree-paypal-button-strategy.ts]: @animesh1987 i have updated this point

**Review by animesh1987 — APPROVED** [review]
